### PR TITLE
[WIP] Formalize what the IO role actually does and introduce IO::NativeDescriptor

### PR DESCRIPTION
**Pull request by Kaiepi**
Originally, this was only supposed to add support for using file descriptors/`HANDLE`s with `IO::Handle`. This alone wasn't quite good enough since `IO::Socket::INET` and `IO::Socket::Async` will also need to be able to support creating sockets given a file descriptor or path once I implement sending file descriptors over sockets as part of my networking grant work, and I want to be able to just pass an `IO` or `IO::Handle` instance to `IO::Socket::INET.new`/`IO::Socket::Async.new` and have it work without hardcoding behaviour specific to file descriptors or paths in v6.e. This requires three different changes in particular.

This starts out by formalizing what `IO` actually is: a role done by classes describing files in some way. Methods that were originally specific to `IO::Path` and `IO::Special` but are generic file operations are now stubbed in `IO`.

Second, this introduces `IO::NativeDescriptor`, an abstraction over file descriptors (`HANDLE`s on Windows). It can be used in the same way `IO::Path` and `IO::Special` are.

Finally, this makes whatever uses `IO` as a type more generic in order to make it possible to create custom file classes. `IO::Handle` no longer hardcodes logic for how `IO`-typed variables should be handled depending on their type in a way that makes that impossible; that is now implemented in the types themselves.

All of this is done while aiming to preserve compatibility with code that already uses `IO` as a type. I'll need to grep the ecosystem for cases where people specify `IO` as a type when they really mean `IO::Path` before this can be merged, as that's the only code that I know of that this may break.

This is unfinished and hasn't been fully tested past the first commit. There are still several nqp ops for dealing with native descriptors that haven't been implemented yet, as well as changes to RI necessary for `IO::NativeDescriptor`'s stat methods to work. Roast will need tests for `IO::NativeDescriptor` and the docs will need to be updated. This will need testing on Windows once it's finished.

### CODE REVIEW

File: src/core/IO/Handle.pm6 [hunk 1]
```diff
@@ -1,8 +1,7 @@
-my class IO::Path { ... }
 my class Proc { ... }
 
 my class IO::Handle {
-    has $.path;
```
Comment (ugexe) [inline src/core/IO/Handle.pm6]: I disagree with this rename. I use `IO::Handle` based abstractions for http, pipes, etc which all have a path/uri representation but are not files. It also seems to add some needless complexity to the constructor which now can seemingly take both `path` and `file`.

File: src/core/IO/NativeDescriptor.pm6 [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    multi method perl(IO::NativeDescriptor:D: --> Str:D) {
+        nqp::p6box_s(
+          nqp::concat(
+            nqp::concat(self.^name, '.new('),
+            nqp::concat(nqp::coerce_is(nqp::unbox_i(self)), ')')
+          )
+        )
+    }
+
+    multi method e       (IO::NativeDescriptor:D: --> Bool)    {
```
Comment (ugexe) [inline src/core/IO/NativeDescriptor.pm6]: This indentation style isn't used anywhere else. Adding any new method that has a longer name will require breaking the `blame` for all methods when they need to be reindented.

File: src/core/IO/Socket/Async.pm6 [hunk 3]
```diff
@@ -208,8 +208,8 @@ my class IO::Socket::Async {
             self.bless(:&on-close, :$VMIO-tobe, :$socket-host, :$socket-port);
         }
 
-        method native-descriptor(--> Int) {
-            nqp::filenofh(await $!VMIO-tobe)
+        method native-descriptor(--> IO::NativeDescriptor) {
```
Comment (ugexe) [inline src/core/IO/Socket/Async.pm6]: This seems like a fairly big breaking change

File: src/core/IO/Special.pm6 [hunk 4]
```diff
@@ -1,45 +1,70 @@
...
+          )
+        )
+    }
+
+    multi method e       (IO::Special:D: --> True)    {         }
+    multi method d       (IO::Special:D: --> False)   {         }
+    multi method f       (IO::Special:D: --> False)   {         }
+    multi method s       (IO::Special:D: --> 0)       {         }
+    multi method l       (IO::Special:D: --> False)   {         }
+    multi method r       (IO::Special:D: --> Bool)    {
```
Comment (ugexe) [inline src/core/IO/Special.pm6]: This is supposed to return `Bool:D`, not `Bool`

File: src/core/IO/Special.pm6 [hunk 5]
```diff
@@ -1,45 +1,70 @@
...
-    method accessed(IO::Special:D: --> Instant) { Instant }
-    method changed( IO::Special:D: --> Instant) { Instant }
-    method mode(IO::Special:D: --> Nil) { }
+    multi method x       (IO::Special:D: --> False)   {         }
+    multi method rw      (IO::Special:D: --> False)   {         }
+    multi method rwx     (IO::Special:D: --> False)   {         }
+    multi method modified(IO::Special:D: --> Instant) { Instant }
+    multi method accessed(IO::Special:D: --> Instant) { Instant }
+    multi method changed (IO::Special:D: --> Instant) { Instant }
+    multi method mode    (IO::Special:D: --> IntStr)  { IntStr  }
```
Comment (ugexe) [inline src/core/IO/Special.pm6]: This returned `Nil` before. Not its returning `IntStr`?

File: src/core/Int.pm6 [hunk 6]
```diff
@@ -46,6 +47,10 @@ my class Int does Real { # declared in BOOTSTRAP
 
     method Capture() { die X::Cannot::Capture.new: :what(self) }
 
+    method IO(Int:D: --> IO::NativeDescriptor:D) {
```
Comment (ugexe) [inline src/core/Int.pm6]: This feels pretty wrong

File: src/core/io_operators.pm6 [hunk 7]
```diff
@@ -81,7 +81,7 @@ multi sub dir(IO::Path:D $path, |c) { $path.dir(|c) }
 multi sub dir(IO()       $path, |c) { $path.dir(|c) }
 
 proto sub open($, |) {*}
-multi sub open(IO() $path, |c) { IO::Handle.new(:$path).open(|c) }
+multi sub open(IO() $file, |c) { IO::Handle.new(:$file).open(|c) }
```
Comment (ugexe) [inline src/core/io_operators.pm6]: Again, I disagree with the renaming to `$file` from `$path` here.

**Comment by ugexe** [comment]
There are a *lot* of breaking changes in here.

**Comment by Kaiepi** [comment]
Most of them are mistakes from while I was refactoring. `Int.IO` exists was because I was forgetting `Cool` already had an `IO` method, and I wasn't considering stuff like URLs when I renamed `IO::Handle.path` to `IO::Handle.file`. I was hoping I could keep returning `IO::NativeDescriptor` from `.native-descriptor` since it's a subclass of `Int`, but after thinking for a while you're right, that would break stuff that depends on it being an `Int` specifically.

File: src/core/IO/NativeDescriptor.pm6 (hunk 2 above)
Comment (AlexDaniel) [reply src/core/IO/NativeDescriptor.pm6]: I love vertical alignment but not as much :) Looks weird indeed.

**Comment by lizmat** [comment]
Is this PR still worthwhile salvaging with the conflicts?  Or can it be closed?

**Comment by Kaiepi** [comment]
It'll still be a while yet before I can come back to this.. this can be closed for now.

**Pull request closed without merging**